File: apps/backend/src/todos/todo.service.ts
import {
  Injectable,
  NotFoundException,
  ForbiddenException,
} from "@nestjs/common";
import { TodoRepository } from "./todo.repository";
import { TodoEntity } from "./todo.entity";
import { CreateTodoDto } from "./dto/create-todo.dto";
import { UpdateTodoDto } from "./dto/update-todo.dto";
import { TodoCategoryDto } from "./dto/todo-category.dto";
import { TodoItem, TodoStats, TodoCategory } from "@calendar-todo/shared-types";
import { UserSettingsService } from "../user-settings/user-settings.service";
import { subDays } from "date-fns";

@Injectable()
export class TodoService {
  constructor(
    private readonly todoRepository: TodoRepository,
    private readonly userSettingsService: UserSettingsService,
  ) {}

  private convertCategoryDtoToCategory(
    categoryDto: TodoCategoryDto,
  ): TodoCategory {
    return {
      id: categoryDto.id,
      name: categoryDto.name,
      color: categoryDto.color,
      icon: categoryDto.icon,
      createdAt: new Date(categoryDto.createdAt),
      order: categoryDto.order,
    };
  }

  async create(
    createTodoDto: CreateTodoDto,
    userId: string,
  ): Promise<TodoItem> {
    const category = this.convertCategoryDtoToCategory(createTodoDto.category);

    // Verify category belongs to user
    const userCategory = await this.userSettingsService.getCategoryById(
      userId,
      category.id,
    );
    if (!userCategory) {
      throw new NotFoundException(
        "Category not found or does not belong to user",
      );
    }

    const todoData = {
      title: createTodoDto.title,
      description: createTodoDto.description,
      priority: createTodoDto.priority || "medium",
      categoryId: category.id,
      todoType: createTodoDto.todoType || "event", // 기본값을 'event'로 설정
      dueDate: new Date(createTodoDto.date),
      userId,
    };

    const todo = await this.todoRepository.create(todoData);
    return todo.toTodoItem(userCategory);
  }

  async findAll(
    userId: string,
    startDate?: string,
    endDate?: string,
    categoryId?: string,
    completed?: boolean,
  ): Promise<TodoItem[]> {
    let todos: TodoEntity[];

    if (startDate && endDate) {
      todos = await this.todoRepository.findByUserIdAndDateRange(
        userId,
        new Date(startDate),
        new Date(endDate),
      );
    } else if (categoryId) {
      todos = await this.todoRepository.findByUserIdAndCategory(
        userId,
        categoryId,
      );
    } else if (completed !== undefined) {
      todos = await this.todoRepository.findByUserIdAndCompleted(
        userId,
        completed,
      );
    } else {
      todos = await this.todoRepository.findByUserId(userId);
    }

    // Get user categories to resolve category information
    const userCategories =
      await this.userSettingsService.getUserCategories(userId);
    const categoryMap = new Map(userCategories.map((cat) => [cat.id, cat]));

    return todos.map((todo) => {
      const category = categoryMap.get(todo.categoryId);
      return todo.toTodoItem(category);
    });
  }

  async findOne(id: string, userId: string): Promise<TodoItem> {
    const todo = await this.todoRepository.findById(id);
    if (!todo) {
      throw new NotFoundException("할일을 찾을 수 없습니다");
    }

    if (todo.userId !== userId) {
      throw new ForbiddenException("해당 할일에 접근할 권한이 없습니다");
    }

    const category = await this.userSettingsService.getCategoryById(
      userId,
      todo.categoryId,
    );
    return todo.toTodoItem(category || undefined);
  }

  async update(
    id: string,
    updateTodoDto: UpdateTodoDto,
    userId: string,
  ): Promise<TodoItem> {
    const todo = await this.todoRepository.findById(id);
    if (!todo) {
      throw new NotFoundException("할일을 찾을 수 없습니다");
    }

    if (todo.userId !== userId) {
      throw new ForbiddenException("해당 할일을 수정할 권한이 없습니다");
    }

    const updateData: Partial<TodoEntity> = {};

    if (updateTodoDto.title !== undefined)
      updateData.title = updateTodoDto.title;
    if (updateTodoDto.description !== undefined)
      updateData.description = updateTodoDto.description;
    if (updateTodoDto.completed !== undefined)
      updateData.completed = updateTodoDto.completed;
    if (updateTodoDto.priority !== undefined)
      updateData.priority = updateTodoDto.priority;
    if (updateTodoDto.category !== undefined) {
      const category = this.convertCategoryDtoToCategory(
        updateTodoDto.category,
      );

      // Verify category belongs to user
      const userCategory = await this.userSettingsService.getCategoryById(
        userId,
        category.id,
      );
      if (!userCategory) {
        throw new NotFoundException(
          "Category not found or does not belong to user",
        );
      }

      updateData.categoryId = category.id;
    }
    if (updateTodoDto.date !== undefined)
      updateData.dueDate = new Date(updateTodoDto.date);
    if (updateTodoDto.todoType !== undefined)
      updateData.todoType = updateTodoDto.todoType;

    const updatedTodo = await this.todoRepository.update(id, updateData);
    const category = await this.userSettingsService.getCategoryById(
      userId,
      updatedTodo!.categoryId,
    );
    return updatedTodo!.toTodoItem(category || undefined);
  }

  async remove(
    id: string,
    userId: string,
  ): Promise<{ success: boolean; deletedId: string }> {
    const todo = await this.todoRepository.findById(id);
    if (!todo) {
      throw new NotFoundException("할일을 찾을 수 없습니다");
    }

    if (todo.userId !== userId) {
      throw new ForbiddenException("해당 할일을 삭제할 권한이 없습니다");
    }

    const success = await this.todoRepository.delete(id);
    return { success, deletedId: id };
  }

  async toggle(id: string, userId: string): Promise<TodoItem> {
    const todo = await this.todoRepository.findById(id);
    if (!todo) {
      throw new NotFoundException("할일을 찾을 수 없습니다");
    }

    if (todo.userId !== userId) {
      throw new ForbiddenException("해당 할일을 수정할 권한이 없습니다");
    }

    const updatedTodo = await this.todoRepository.toggle(id);
    const category = await this.userSettingsService.getCategoryById(
      userId,
      updatedTodo!.categoryId,
    );
    return updatedTodo!.toTodoItem(category || undefined);
  }

  async getStats(userId: string): Promise<TodoStats> {
    // Use efficient Redis counters instead of loading all todos
    const stats = await this.todoRepository.getStatsForUser(userId);

    // For recent completions, query only completed todos and filter by date
    // This is more efficient than loading all todos
    const sevenDaysAgo = subDays(new Date(), 7);
    const completedTodos = await this.todoRepository.findByUserIdAndCompleted(
      userId,
      true,
    );
    const recentCompletions = completedTodos.filter(
      (todo) => todo.updatedAt >= sevenDaysAgo,
    ).length;

    const completionRate =
      stats.total > 0 ? Math.round((stats.completed / stats.total) * 100) : 0;

    // Get all todos to calculate type-based statistics
    const allTodos = await this.todoRepository.findByUserId(userId);

    // Calculate statistics by todo type
    const eventTodos = allTodos.filter((todo) => todo.todoType === "event");
    const taskTodos = allTodos.filter((todo) => todo.todoType === "task");

    const eventStats = {
      total: eventTodos.length,
      completed: eventTodos.filter((todo) => todo.completed).length,
      incomplete: eventTodos.filter((todo) => !todo.completed).length,
    };

    const taskStats = {
      total: taskTodos.length,
      completed: taskTodos.filter((todo) => todo.completed).length,
      incomplete: taskTodos.filter((todo) => !todo.completed).length,
    };

    return {
      total: stats.total,
      completed: stats.completed,
      incomplete: stats.incomplete,
      completionRate,
      recentCompletions,
      byType: {
        event: eventStats,
        task: taskStats,
      },
    };
  }

  async removeAllByUserId(userId: string): Promise<number> {
    const todos = await this.todoRepository.findByUserId(userId);
    await this.todoRepository.deleteByUserId(userId);
    return todos.length;
  }

  async updateCategoryForUser(
    userId: string,
    oldCategoryId: string,
    newCategoryId: string,
  ): Promise<number> {
    return await this.todoRepository.updateCategoryForUser(
      userId,
      oldCategoryId,
      newCategoryId,
    );
  }

  async bulkCreate(
    todos: Omit<TodoItem, "id">[],
    userId: string,
  ): Promise<TodoItem[]> {
    const createdTodos: TodoItem[] = [];

    for (const todoData of todos) {
      const categoryDto: TodoCategoryDto = {
        id: todoData.category.id,
        name: todoData.category.name,
        color: todoData.category.color,
        icon: todoData.category.icon,
        createdAt: todoData.category.createdAt.toISOString(),
        order: todoData.category.order,
      };

      const createDto: CreateTodoDto = {
        title: todoData.title,
        category: categoryDto,
        date: todoData.date.toISOString(),
        priority: "medium",
      };

      const todo = await this.create(createDto, userId);
      if (todoData.completed) {
        await this.update(todo.id, { completed: true }, userId);
      }
      createdTodos.push(todo);
    }

    return createdTodos;
  }

  /**
   * 오늘 이전 날짜의 미완료 작업들을 오늘로 이동
   */
  async moveTasksToNextDay(
    userId: string,
  ): Promise<{ movedCount: number; movedTaskIds: string[] }> {
    const today = new Date();
    today.setHours(0, 0, 0, 0);

    // 이동이 필요한 작업들 조회
    const tasksToMove = await this.getTasksDueForMove(userId);

    if (tasksToMove.length === 0) {
      return { movedCount: 0, movedTaskIds: [] };
    }

    let movedCount = 0;
    const movedTaskIds: string[] = [];

    // 각 작업을 오늘로 이동
    for (const task of tasksToMove) {
      try {
        await this.todoRepository.update(task.id, {
          dueDate: today,
          updatedAt: new Date(),
        });
        movedCount++;
        movedTaskIds.push(task.id);
      } catch (error) {
        console.error(`작업 ${task.id} 이동 실패:`, error);
      }
    }

    return { movedCount, movedTaskIds };
  }

  /**
   * 이동이 필요한 작업들 조회
   */
  async getTasksDueForMove(userId: string): Promise<TodoEntity[]> {
    const today = new Date();
    today.setHours(0, 0, 0, 0);

    // 사용자의 모든 할 일 조회
    const allTodos = await this.todoRepository.findByUserId(userId);

    // 이동 조건에 맞는 작업들 필터링
    return allTodos.filter((todo) => {
      return (
        todo.todoType === "task" && // 작업 타입
        !todo.completed && // 미완료
        todo.dueDate < today // 오늘 이전 날짜
      );
    });
  }

  /**
   * 특정 사용자의 이동 대상 작업 개수 조회
   */
  async getTasksMoveCount(userId: string): Promise<number> {
    const tasksToMove = await this.getTasksDueForMove(userId);
    return tasksToMove.length;
  }
}

File: apps/backend-python/app/services/todo_service.py
"""
Todo service for todo management operations.
"""
from typing import List, Optional, Dict, Any
from datetime import datetime, timedelta

from app.models.todo import Todo, TodoType
from app.models.category import Category
from app.repositories.todo_repository import TodoRepository
from app.repositories.category_repository import CategoryRepository
from app.schemas.todo import TodoCreate, TodoUpdate, TodoResponse, TodoStats


class TodoService:
    """Service for todo management operations."""
    
    def __init__(
        self,
        todo_repository: TodoRepository,
        category_repository: CategoryRepository,
    ):
        self.todo_repository = todo_repository
        self.category_repository = category_repository
    
    async def create_todo(
        self, user_id: str, todo_data: TodoCreate
    ) -> TodoResponse:
        """Create a new todo."""
        # Verify category belongs to user
        category = await self.category_repository.find_by_id(
            user_id, todo_data.category_id
        )
        if not category:
            raise ValueError("Category not found or does not belong to user")
        
        # Create todo
        todo = Todo(
            id=Todo.generate_id(),
            title=todo_data.title,
            description=todo_data.description,
            date=todo_data.date,
            completed=False,
            completed_at=None,
            todo_type=todo_data.todo_type or TodoType.EVENT,
            priority=todo_data.priority,
            category_id=todo_data.category_id,
            user_id=user_id,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
        )
        
        await self.todo_repository.save(todo)
        
        return self._to_todo_response(todo, category)
    
    async def get_todos(
        self,
        user_id: str,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        category_id: Optional[str] = None,
        completed: Optional[bool] = None,
    ) -> List[TodoResponse]:
        """Get todos with filters."""
        if start_date and end_date:
            todos = await self.todo_repository.find_by_date_range(
                user_id, start_date, end_date
            )
        elif category_id:
            todos = await self.todo_repository.find_by_category(user_id, category_id)
        elif completed is not None:
            if completed:
                todos = await self.todo_repository.find_completed_for_user(user_id)
            else:
                todos = await self.todo_repository.find_incomplete_for_user(user_id)
        else:
            todos = await self.todo_repository.find_all_for_user(user_id)
        
        # Get categories for todos
        category_ids = list(set(todo.category_id for todo in todos))
        categories = []
        for cat_id in category_ids:
            cat = await self.category_repository.find_by_id(user_id, cat_id)
            if cat:
                categories.append(cat)
        
        category_map = {cat.id: cat for cat in categories}
        
        return [
            self._to_todo_response(todo, category_map.get(todo.category_id))
            for todo in todos
        ]
    
    async def get_todo_by_id(self, user_id: str, todo_id: str) -> TodoResponse:
        """Get a specific todo."""
        todo = await self.todo_repository.find_by_id(user_id, todo_id)
        if not todo:
            raise ValueError("Todo not found")
        
        category = await self.category_repository.find_by_id(user_id, todo.category_id)
        return self._to_todo_response(todo, category)
    
    async def update_todo(
        self, user_id: str, todo_id: str, update_data: TodoUpdate
    ) -> TodoResponse:
        """Update a todo."""
        todo = await self.todo_repository.find_by_id(user_id, todo_id)
        if not todo:
            raise ValueError("Todo not found")
        
        # Verify new category if provided
        if update_data.category_id:
            category = await self.category_repository.find_by_id(
                user_id, update_data.category_id
            )
            if not category:
                raise ValueError("Category not found or does not belong to user")
        
        # Update fields
        if update_data.title is not None:
            todo.title = update_data.title
        if update_data.description is not None:
            todo.description = update_data.description
        if update_data.date is not None:
            todo.date = update_data.date
        if update_data.priority is not None:
            todo.priority = update_data.priority
        if update_data.category_id is not None:
            todo.category_id = update_data.category_id
        if update_data.todo_type is not None:
            todo.todo_type = update_data.todo_type
        if update_data.completed is not None:
            todo.completed = update_data.completed
            todo.completed_at = datetime.utcnow() if update_data.completed else None
        
        todo.updated_at = datetime.utcnow()
        
        await self.todo_repository.save(todo)
        
        category = await self.category_repository.find_by_id(user_id, todo.category_id)
        return self._to_todo_response(todo, category)
    
    async def toggle_todo(self, user_id: str, todo_id: str) -> TodoResponse:
        """Toggle todo completion status."""
        updated = await self.todo_repository.update_completion_status(
            user_id, todo_id, None  # Toggle
        )
        if not updated:
            raise ValueError("Todo not found")
        
        category = await self.category_repository.find_by_id(
            user_id, updated.category_id
        )
        return self._to_todo_response(updated, category)
    
    async def delete_todo(self, user_id: str, todo_id: str) -> Dict[str, Any]:
        """Delete a todo."""
        deleted = await self.todo_repository.delete(user_id, todo_id)
        if not deleted:
            raise ValueError("Todo not found")
        
        return {"success": True, "deletedId": todo_id}
    
    async def delete_all_todos(self, user_id: str) -> Dict[str, Any]:
        """Delete all todos for a user."""
        count = await self.todo_repository.delete_all_for_user(user_id)
        return {"success": True, "deletedCount": count}
    
    async def get_stats(self, user_id: str) -> TodoStats:
        """Get todo statistics."""
        # Get basic stats
        stats = await self.todo_repository.get_stats_for_user(user_id)
        
        # Calculate recent completions (last 7 days)
        seven_days_ago = datetime.utcnow() - timedelta(days=7)
        completed_todos = await self.todo_repository.find_completed_for_user(user_id)
        recent_completions = sum(
            1 for todo in completed_todos
            if todo.completed_at and todo.completed_at >= seven_days_ago
        )
        
        # Calculate completion rate
        completion_rate = (
            round((stats["completed"] / stats["total"]) * 100)
            if stats["total"] > 0
            else 0
        )
        
        # Get todos by type
        all_todos = await self.todo_repository.find_all_for_user(user_id)
        event_todos = [t for t in all_todos if t.todo_type == TodoType.EVENT]
        task_todos = [t for t in all_todos if t.todo_type == TodoType.TASK]
        
        return TodoStats(
            total=stats["total"],
            completed=stats["completed"],
            incomplete=stats["incomplete"],
            completion_rate=completion_rate,
            recent_completions=recent_completions,
            event_stats={
                "total": len(event_todos),
                "completed": sum(1 for t in event_todos if t.completed),
                "incomplete": sum(1 for t in event_todos if not t.completed),
            },
            task_stats={
                "total": len(task_todos),
                "completed": sum(1 for t in task_todos if t.completed),
                "incomplete": sum(1 for t in task_todos if not t.completed),
            },
        )
    
    async def move_tasks(
        self, user_id: str, from_date: datetime, to_date: datetime
    ) -> List[TodoResponse]:
        """Move incomplete tasks from one date to another."""
        # Get incomplete tasks on the from_date
        todos = await self.todo_repository.find_by_date_range(
            user_id, from_date, from_date
        )
        
        tasks_to_move = [
            t for t in todos
            if t.todo_type == TodoType.TASK and not t.completed
        ]
        
        moved_tasks = []
        for task in tasks_to_move:
            task.date = to_date
            task.updated_at = datetime.utcnow()
            await self.todo_repository.save(task)
            
            category = await self.category_repository.find_by_id(
                user_id, task.category_id
            )
            moved_tasks.append(self._to_todo_response(task, category))
        
        return moved_tasks
    
    async def get_tasks_due_for_movement(
        self, user_id: str, before_date: datetime
    ) -> List[TodoResponse]:
        """Get incomplete tasks that are due for movement."""
        todos = await self.todo_repository.find_all_for_user(user_id)
        
        tasks_due = [
            t for t in todos
            if t.todo_type == TodoType.TASK
            and not t.completed
            and t.date < before_date
        ]
        
        # Get categories
        category_ids = list(set(t.category_id for t in tasks_due))
        categories = []
        for cat_id in category_ids:
            cat = await self.category_repository.find_by_id(user_id, cat_id)
            if cat:
                categories.append(cat)
        
        category_map = {cat.id: cat for cat in categories}
        
        return [
            self._to_todo_response(task, category_map.get(task.category_id))
            for task in tasks_due
        ]
    
    def _to_todo_response(
        self, todo: Todo, category: Optional[Category]
    ) -> TodoResponse:
        """Convert todo model to response."""
        return TodoResponse(
            id=todo.id,
            title=todo.title,
            description=todo.description,
            date=todo.date.isoformat(),
            completed=todo.completed,
            completed_at=todo.completed_at.isoformat() if todo.completed_at else None,
            priority=todo.priority,
            todo_type=todo.todo_type,
            category={
                "id": category.id,
                "name": category.name,
                "color": category.color,
                "icon": category.icon,
                "created_at": category.created_at.isoformat(),
                "order": category.order,
            } if category else None,
            created_at=todo.created_at.isoformat(),
            updated_at=todo.updated_at.isoformat(),
        )